lib/pos_blockchain.go

```diff
@@ -481,9 +482,164 @@ func (bc *Blockchain) pruneUncommittedBlocks(desoBlock *MsgDeSoBlock) error {
 // commitGrandparents commits the grandparent of the block if possible.
 // Specifically, this updates the CommittedBlockStatus of its grandparent
 // and flushes the view after connecting the grandparent block to the DB.
-func (bc *Blockchain) commitGrandparents(desoBlock *MsgDeSoBlock) error {
-	// TODO: Implement me.
-	return errors.New("IMPLEMENT ME")
+func (bc *Blockchain) commitGrandparents() error {
+	currentBlock := bc.GetBestChainTip()
+	var canCommit bool
+	var blockToCommit *BlockHash
+	for !canCommit {
+		// If the grandparent block is committed, then we can commit it.
+		// Otherwise, we can't commit it.
+		var committedBlockSeen bool
+		blockToCommit, canCommit, committedBlockSeen = bc.canCommitGrandparent(currentBlock)
+		// If we've seen a committed block, then we can't commit anything new.
+		if committedBlockSeen {
+			return nil
+		}
+		if !canCommit {
+			var exists bool
+			currentBlock, exists = bc.bestChainMap[*currentBlock.Header.PrevBlockHash]
+			if !exists {
+				return errors.Errorf("commitGrandparents: Block %v not found in best chain map", currentBlock.Hash.String())
+			}
+		}
+	}
+	// Find all uncommitted ancestors of block to commit
+	_, idx := bc.getHighestCommittedBlock()
+	if idx == -1 {
+		// This is an edge case we'll never hit in practice since all the PoW blocks
+		// are committed.
+		return errors.New("commitGrandparents: No committed blocks found")
+	}
+	uncommittedAncestors := []*BlockNode{}
+	for ii := idx + 1; ii < len(bc.bestChain); ii++ {
+		uncommittedAncestors = append(uncommittedAncestors, bc.bestChain[ii])
+		if bc.bestChain[ii].Hash.IsEqual(blockToCommit) {
+			break
+		}
+	}
+	for ii := 0; ii < len(uncommittedAncestors); ii++ {
+		if err := bc.commitBlock(uncommittedAncestors[ii].Hash); err != nil {
+			return errors.Wrapf(err, "commitGrandparents: Problem committing block %v", uncommittedAncestors[ii].Hash.String())
+		}
+	}
+	return nil
+}
+
+// canCommitGrandparent determines if the grandparent of the current block can be committed.
+// The grandparent can be committed if there exists a direct parent-child relationship
+// between the grandparent and parent of the new block, meaning the grandparent and parent
+// are proposed in consecutive views, and the "parent" is an ancestor of the incoming block (not necessarily consecutive views).
+// Additionally, the grandparent must not already be committed.
+func (bc *Blockchain) canCommitGrandparent(currentBlock *BlockNode) (_grandparentBlockHash *BlockHash, _canCommit bool, _committedBlockSeen bool) {
```

Comment:
What is the _committedBlockSeen return param used for? IMO, it's sufficient to just use the _canCommit param to define whether the grand can or cannot be committed.

Reply:
This was only necessary when I was looping over ancestors. If we hit a "committed" block then we could stop the search. This is no longer necessary and will be removed.